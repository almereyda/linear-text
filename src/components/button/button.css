/* Buttonish classes are shared by other button-like components. */
.buttonish {
  /* Dimensions include border and padding. */
  box-sizing: border-box;

  min-width: var(--min-press-size);
  min-height: var(--min-press-size);

  /* Require buttons to be spaced apart so that misses don't interact with
     something else. */
  margin-top: var(--button-margin, 2px);
  margin-right: var(--button-margin, 2px);
  margin-bottom: var(--button-margin, 2px);
  margin-left: var(--button-margin, 2px);

  /* Add some spacing between the button content and the perimeter but account
     for border-width's contribution to --min-press-size. */
  padding-top: 5.5px;
  padding-right: 5.5px;
  padding-bottom: 5.5px;
  padding-left: 5.5px;

  /* Indicate button interactivity with a hand icon. */
  cursor: pointer;

  /* Buttons slightly extrude. */
  border-width: 0.5px;
  border-style: outset;
  border-color: var(--color-dark-grey);

  /* Specify the background. */
  background-color: var(--color-grey);

  /* Inherit font. */
  font-family: inherit;
  font-size: inherit;
  font-stretch: inherit;
  font-style: inherit;
  font-variant: inherit;
  font-weight: inherit;
  line-height: inherit;

  /* Round corners. */
  border-radius: var(--border-radius);

  /* Don't allow content highlights. */
  user-select: none;

  /* Center the content. */
  text-align: center;

  /* For pseudo-element positioning. */
  position: relative;

  /* Button-likes may specify different display types. */
  display: inline-flex;
}

.buttonish:active,
/* --active state can be held too. */
.buttonish--active {
  border-style: inset;
}

/* Tooltip. */
.buttonish[data-title]::after {
  /* Set the child to the value of the data-title attribute. */
  content: attr(data-title);

  font-size: var(--font-size-tooltip);

  /* Position to the bottom-left of the buttonish and remove from layout
     flow. */
  position: absolute;
  top: 100%;
  left: 0;

  /* Don't wrap the text. Width defaults to buttonish. */
  white-space: nowrap;

  /* Raise. */
  z-index: var(--z-index-tooltip);

  /* Use a dark theme for tooltips so they contrast from the base UI. */
  background-color: var(--color-darkest-grey);
  color: var(--color-white);

  /* Make card-like. */
  box-shadow: var(--box-shadow);
  border-radius: var(--border-radius);
  margin-top: var(--card-margin);
  margin-right: var(--card-margin);
  margin-bottom: var(--card-margin);
  margin-left: var(--card-margin);
  padding-top: var(--card-padding);
  padding-right: var(--card-padding);
  padding-bottom: var(--card-padding);
  padding-left: var(--card-padding);

  /* Hide by default. */
  visibility: hidden;

  /* Reset the tooltip show delay promptly. */
  transition-delay: 50ms;

  /* Ignore interactions. */
  pointer-events: none;
}

.buttonish[data-title]:hover::after {
  /* Show after a long delay. */
  visibility: visible;
  transition-delay: 2s;
}
