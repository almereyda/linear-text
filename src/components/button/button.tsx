import {MouseEventHandler, ReactNode, useMemo} from 'react'

import './button.css'

export type ButtonProps = Readonly<{
  accessKey?: string
  active?: boolean
  children: ReactNode
  'data-testid'?: string
  disabled?: boolean
  onClick: MouseEventHandler<HTMLButtonElement>
  title: string
}>

export function Button({
  accessKey,
  children,
  'data-testid': testID,
  disabled,
  active,
  onClick,
  title
}: ButtonProps) {
  const className = useMemo(
    () => `buttonish ${active ? 'buttonish--active' : ''}`,
    [active]
  )
  return (
    <button
      accessKey={accessKey ?? ''}
      className={className}
      data-testid={testID}
      data-title={title}
      disabled={disabled ?? false}
      onClick={onClick}
    >
      {children}
    </button>
  )
}
