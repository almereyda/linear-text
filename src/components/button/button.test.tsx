import {Button} from './button'
import {render, fireEvent, screen} from '@testing-library/react'

test('matches snapshot', () => {
  render(
    <Button data-testid='button' onClick={() => {}} title='Title'>
      Label
    </Button>
  )
  const button = screen.getByTestId('button')
  expect(button).toMatchSnapshot()
})

test('renders children and title', () => {
  render(
    <Button data-testid='button' onClick={() => {}} title='Title'>
      Label
    </Button>
  )
  const button = screen.getByTestId('button')
  expect(button).toHaveTextContent('Label')
  expect(button).toHaveAttribute('title', 'Title')
})

test('reports clicks', () => {
  const onClick = jest.fn()
  render(
    <Button data-testid='button' onClick={onClick} title='Title'>
      Label
    </Button>
  )
  fireEvent.click(screen.getByTestId('button'))
  expect(onClick).toHaveBeenCalledTimes(1)
})
