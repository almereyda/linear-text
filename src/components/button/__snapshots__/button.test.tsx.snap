// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<button
  accesskey=""
  class="buttonish "
  data-testid="button"
  data-title="Title"
>
  Label
</button>
`;
